### PR TITLE
remote integration test

### LINKED ISSUE #535: Connect to testnet in CI to guard against incompatibility

### PR DESCRIPTION
**Pull request by Trivo25**
Connects a basic zkApp contract to the berkeley testnet 

TODO: same thing in the mina repo

SnarkyJS part for https://github.com/MinaProtocol/mina/issues/11248

closes #535

### CODE REVIEW

**Comment by mitschabaude** [comment]
cool! arguably it's a separate issue, which we had already tracked: #535

File: src/lib/mina.ts [hunk 1]
```diff
@@ -675,7 +675,7 @@ function Network(graphqlEndpoint: string): Mina {
             } else if (maxAttempts && attempts === maxAttempts) {
               return reject(
                 new Error(
-                  `Exceeded max attempts. TransactionId: ${txId}, attempts: ${attempts}`
+                  `Exceeded max attempts. TransactionId: ${txId}, attempts: ${attempts}, last received status: ${res}`
```
Comment (nicc) [inline src/lib/mina.ts:678]: Nice

**Review by nicc — APPROVED** [review]
TL;DR: looks good to me. 

Just writing out what I understand to be happening here...
- init the HelloWorld app
- check if it’s deployed by fetching some state
- deploy it if not
- check that we can fetch accounts
- check that we can update some state
- check that the state has been updated by fetching it again

We’re not making test assertions here, just running a JS script that will error if there’s a problem. So we’re not running any test frameworks and will rely on the exit code of the script. Seems like a very reasonable choice in this context.

The only feedback that comes to mind is that `faucet()` could be a useful utility for other tests and I wonder if its existence will be known to future contributors. Two questions: do you think it's reuasable? And is there a place for test utils?

File: src/examples/zkapps/hello_world/run_berkeley.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+  let transaction = await Mina.transaction(
+    { feePayerKey, fee: transactionFee },
+    () => {
+      AccountUpdate.fundNewAccount(feePayerKey);
+      zkapp.deploy({ zkappKey, verificationKey });
+    }
+  );
+
+  console.log('Sending the transaction..');
+  await (await transaction.send()).wait();
```
Comment (mitschabaude) [inline src/examples/zkapps/hello_world/run_berkeley.ts:59]: wow, really powerful to have wait() here

**Comment by Trivo25** [comment]
> The only feedback that comes to mind is that faucet() could be a useful utility for other tests and I wonder if its existence will be known to future contributors. Two questions: do you think it's reuasable? And is there a place for test utils?

Very good point. I feel like we can now create a dedicated test lib file, as there are a handful of helper functions we use (tictoc, profiler, and now the faucet/waiting for tx). Will add that!

**Review by mitschabaude — APPROVED** [review]
awesome! I hope it passes :D

File: src/examples/zkapps/hello_world/run_berkeley.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+let transactionFee = 100_000_000;
+
+console.log('Compiling smart contract..');
+let { verificationKey } = await HelloWorld.compile();
+
+let zkapp = new HelloWorld(zkappAddress);
+let x = await zkapp.x.fetch();
+let isDeployed = x?.equals(0).not().toBoolean() ?? false;
+
+if (!isDeployed) {
```
Comment (mitschabaude) [inline src/examples/zkapps/hello_world/run_berkeley.ts:47]: nit: we don't need this check here. there's no way the random public key can already have a deployed zkapp on it

**Comment by mitschabaude** [comment]
> QQ: Do we want to introduce a Testing namespace that supplies those helper functions to the outside world?

I'd say no for now, because the testing helpers will use node JS APIs, while snarkyjs has to be web compatible as well.
